lib/charms/prometheus_k8s/v0/prometheus_scrape.py

```diff
@@ -1899,25 +1931,41 @@ def _set_prometheus_data(self, event):
         `MetricsEndpointAggregator`, that Prometheus unit is provided
         with the complete set of existing scrape jobs and alert rules.
         """
-        jobs = []  # list of scrape jobs, one per relation
+        jobs = [] + _type_convert_stored(
+            self._stored.jobs
+        )  # list of scrape jobs, one per relation
         for relation in self.model.relations[self._target_relation]:
             targets = self._get_targets(relation)
             if targets and relation.app:
                 jobs.append(self._static_scrape_job(targets, relation.app.name))
 
-        groups = []  # list of alert rule groups, one group per relation
+        groups = [] + _type_convert_stored(self._stored.alert_rules)  # list of alert rule groups
         for relation in self.model.relations[self._alert_rules_relation]:
             unit_rules = self._get_alert_rules(relation)
             if unit_rules and relation.app:
                 appname = relation.app.name
                 rules = self._label_alert_rules(unit_rules, appname)
-                group = {"name": self._group_name(appname), "rules": rules}
+                group = {"name": self.group_name(appname), "rules": rules}
                 groups.append(group)
 
         event.relation.data[self._charm.app]["scrape_jobs"] = json.dumps(jobs)
         event.relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": groups})
 
-    def _set_target_job_data(self, targets: dict, app_name: str, **kwargs) -> None:
+    def _on_prometheus_targets_changed(self, event):
+        """Update scrape jobs in response to scrape target changes.
+
+        When there is any change in relation data with any scrape
+        target, the Prometheus scrape job, for that specific target is
+        updated.
+        """
+        targets = self._get_targets(event.relation)
+        if not targets:
+            return
```

Comment:
It's not, really.

Since the Aggregator is "fan out", any given `RelationChangedEvent` may or may not actually include target data (it's possible for a `prometheus-target` interface, in reactive-land) to send information which is not, in fact target data.

In that case, nothing we are sending to Prometheus has actually changed, we don't have any updates, so we just `return`. `set_target_job_data` does not updated stored data per-se, but sifts through the current relation data, adds the newly-created job if it's not present, then sets relation data. This is still necessary.